src/ws/BeamSocket.ts

```diff
@@ -0,0 +1,588 @@
+ * @return {Promise}
+ */
+function timeout(delay: number): Promise<void> {
+    return new BeamSocket.Promise<void>((_resolve, reject) => {
+        setTimeout(() => reject(new TimeoutError()), delay);
+    });
+}
+
+function isBrowserWebSocket(socket: any): socket is WebSocket {
+    return !socket.ping;
+}
+
+function isNodeWebSocket(socket: any): socket is NodeWebSocket {
+    return !isBrowserWebSocket(socket);
+}
+
+/**
+ * Wraps a DOM socket with EventEmitter-like syntax.
+ */
+function wrapDOM(socket: WebSocket) {
+    function wrapHandler(event: string, fn: (ev: Event) => void) {
+        return (ev: Event) => {
+            if (event === 'message') {
+                fn((<MessageEvent>ev).data);
+            } else {
+                fn(ev);
+            }
+        };
+    }
+
+    (<any>socket).on = (event: string, listener: (ev: MessageEvent) => void) => {
+        const wrapped = wrapHandler(event, listener);
+        socket.addEventListener(event, wrapped);
+    };
+
+    (<any>socket).once = (event: string, listener: (ev: MessageEvent) => void) => {
+        const wrapped = wrapHandler(event, listener);
+        socket.addEventListener(event, ev => {
+            wrapped(ev);
+            socket.removeEventListener(event, wrapped);
+        });
+    };
+
+    return socket;
+}
+
+export interface IGenericWebSocket {
+    // tslint:disable-next-line: no-misused-new
+    new (address: string): IGenericWebSocket;
+    close() : void;
+    on(ev: string, listener: (arg: any) => void): this;
+    once(ev: string, listener: (arg: any) => void): this;
+    send(data: string): void;
+}
+
+export interface ICallOptions {
+    /**
+     * Set to false if you want a Promise to return for when the event is sent and received by the chat server.
+     */
+    noReply?: boolean;
+    /**
+     * Set to true if you want to force send a event to the server.
+     */
+    force?: boolean;
+
+    /**
+     * Call timeout.
+     */
+    timeout?: number;
+}
+
+/**
+ * Manages a connect to Beam's chat servers.
+ */
+export class BeamSocket extends EventEmitter {
+    private _addressOffset: number;
+    // Spool to store events queued when the connection is lost.
+    private _spool: { data: any, resolve: any }[] = [];
+    private _addresses: string[];
+    // The WebSocket instance we're currently connected with.
+    private ws: IGenericWebSocket;
+    private _pingTimeoutHandle: NodeJS.Timer | number;
+    // Counter of the current number of reconnect retries, and the number of
+    // retries before we reset our reconnect attempts.
+    private _retries: number = 0;
+    private _retryWrap: number = 7; // max 2 minute retry time;
+    private _reconnectTimeout: NodeJS.Timer | number;
+    private _callNo: number;
+    private status: number;
+    private _authpacket: [number, number, string];
+    private _replies: { [key: string]: Reply };
+
+    /**
+     * We've not tried connecting yet
+     */
+    public static IDLE = 0;
```

Comment:
Not really, there are no static enums just yet, don't want to break too much functionality.